Chore: setup medusa + 1 basic fuzz scenario

0xteddybear opened the pull request:
- used --foundry-compile-all to ensure the test contract under `test/properties` is compiled (otherwise it is not compiled and medusa crashes when it can't find it's compiled representation)
- set src,test,script to test/properties/medusa to not waste time compiling contracts that are not required for the medusa campaign

questions
- [ ] should I leave more of the setup to the fuzzer? less?
- [ ] should I allow for more failure modes? e.g. not bounding the amount to be bridged to the available balance or allow for tokens with zero supply (related to the above)
- [ ] should I enable property testing and replace the `assert` in property_* methods by a return? I opted to not do that for consistency (since all other properties are actually checked with asserts)
- [ ] should 'no free mint' be a property?

to figure out what I could so far see of the first item I ran a benchmark (from an empty corpus):

no initial mint, no bound on transfer amount: 176835 calls in 200s, 713 lines covered
initial mint, bound on transfer amount: 146625 calls in 200s, 643 lines covered

(you're not allowed make fun of my 'puter after running this on an ARM chip, btw)

Is offtopic for this PR (will take care of it in the next one)
- actors
- transfers between differente accounts (will probably require^)

CLOSES FUZZ-28
CLOSES FUZZ-29
CLOSES FUZZ-30

0xteddybear committed:
chore: configure medusa with basic supERC20 self-bridging

- used --foundry-compile-all to ensure the test contract under
  `test/properties` is compiled (otherwise it is not compiled and medusa
  crashes when it can't find it's compiled representation)
- set src,test,script to test/properties/medusa to not waste time
  compiling contracts that are not required for the medusa campaign
- used an atomic bridge, which doesnt allow for testing of several of
  the proposed invariants

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/SuperchainERC20Factory.t.sol:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    // this is a stateless check, so halmos will probably supersede it
+    function testContractAddressDependsOnParams(DeployParams memory left, DeployParams memory right) external {
```
on the opUSDC campaign we used this standard for test contracts names:
```suggestion
    function fuzz_contractAddressDependsOnParams(DeployParams memory left, DeployParams memory right) external {
```

I think we should stick to that unless is a blocker for medusa

0xDiscotech commented on packages/contracts-bedrock/medusa.json:
```diff
@@ -0,0 +1,82 @@
+		],
+		"blockNumberDelayMax": 60480,
+		"blockTimestampDelayMax": 604800,
+		"blockGasLimit": 125000000,
```
on Optimism, the block gas limit is 30m
```suggestion
		"blockGasLimit": 30000000,
```
And also, 20m is the limit of the gas that can be deposited from L1 -> L2 for a given block (jic this has some influence on some config)

0xteddybear replied:
should we have:

- block gas limit be realistic for the deployed network (keep in mind some other things won't be, such as all contracts being deployed in the same EVM as opposed to different L2s)
- block gas limit be unrealistic, but compensate for the above and allow more calls in the same block, which could point us in the direction of an otherwise-unreachable bug?

0xDiscotech replied:
If we can try to try the tests on a realistic environment, is better for me. Otherwise, if it becomes an issue or is really complex to handle, we can increase it

0xDiscotech commented on packages/contracts-bedrock/medusa.json:
```diff
@@ -0,0 +1,82 @@
+			"excludeFunctionSignatures": []
+		},
+		"chainConfig": {
+			"codeSizeCheckDisabled": true,
```
Which is the reason of this check? Is this mandatory?
If not, would be nice to enable it since for example, on opUSDC we had some problems with the code size to be deployed

0xteddybear replied:
from `medusa init` :grin: while I of course think it's worthwhile to include a check on the size of contracts to be deployed, the place for that should be as a CI step, and only concerning itself with code under `src/`

Keep in mind that medusa will only compile the medusa fuzz contracts and whatever is imported from there (see newly defined foundry profile), so using this as a check would be ineffective for that

Furthermore, fuzz contracts are liable to have a huge code size since they have lots of methods + deploy other contracts, so their bytecode has to be included as well, so I believe we'll be re-disabling this in the near future if we were to enable it

0xDiscotech replied:
Ah awesome, this check makes complete sense

0xDiscotech reviewed (commented):
Amazing ser 🐻 

Since this is only the setup I guess that's why you didn't write the property on the tests NATSPEC, so no problem with that but we need to include them in the future. Maybe we can leave a TODO.

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol:
```diff
@@ -0,0 +1,151 @@
+        _;
+    }
+
+    function fuzz_DeployNewSupertoken(
```
Missing assertions?

0xteddybear replied:
I can think of an assertion for property 14, but am I missing another one?

0xDiscotech replied:
~~I mean is ok if is not any property and you're only testing that the setup indeed works. But don't you think using a `fuzz_setup` with a comment explaining this makes it more clear?~~

0xDiscotech commented on packages/contracts-bedrock/test/properties/helpers/Utils.sol:
```diff
@@ -0,0 +1,14 @@
+
+import { MockERC20 } from "forge-std/mocks/MockERC20.sol";
+
+contract FuzzERC20 is MockERC20 {
```
I think this is unused

0xteddybear committed:
fix: delete dead code

0xteddybear committed:
test: give the fuzzer a head start

0xteddybear committed:
test: document & implement assertions 22, 23  and 24

0xteddybear committed:
docs: fix properties order

0xteddybear committed:
fix: fixes from self-review

0xteddybear committed:
test: guide the fuzzer a little bit less

previously: initial mint, bound on transfer amount: 146625 calls in 200s
now: no initial mint, no bound on transfer amount: 176835 calls in 200s

it doesn't seem to slow the fuzzer down

0xDiscotech commented on packages/contracts-bedrock/test/properties/SUMMARY.md:
```diff
@@ -0,0 +1,43 @@
+
+- [ ]  [OptimismSuperchainERC20](https://github.com/defi-wonderland/optimism/pull/9/files#diff-810060510a8a9c06dc60cdce6782e5cafd93b638e2557307a68abe694ee86aee)
+- [ ]  [OptimismMintableERC20Factory](https://github.com/defi-wonderland/optimism/blob/develop/packages/contracts-bedrock/src/universal/OptimismMintableERC20Factory.sol)
+- [ ]  [SuperchsupERC20ainERC20](https://github.com/defi-wonderland/optimism/pull/8/files#diff-603fd7d5a0b2c403c0d1eee21d0ee60fb8eb72430169eaac5ec7081e01de96b8) (not yet merged)
```
```suggestion
- [ ]  [SuperchainERC20](https://github.com/defi-wonderland/optimism/pull/8/files#diff-603fd7d5a0b2c403c0d1eee21d0ee60fb8eb72430169eaac5ec7081e01de96b8) (not yet merged)
```

0xDiscotech commented on packages/contracts-bedrock/test/properties/helpers/MockCrossDomainMessenger.t.sol:
```diff
@@ -0,0 +1,52 @@
+    ///////////////////////////////////////////////////
+    //  Helpers for cross-chain interaction mocking  //
+    ///////////////////////////////////////////////////
+    mapping(address => bytes32) public superTokenInitDeploySalts;
```
NIT: named args on the mapping

0xDiscotech commented on packages/contracts-bedrock/test/properties/helpers/MockCrossDomainMessenger.t.sol:
```diff
@@ -0,0 +1,52 @@
+import { OptimismSuperchainERC20 } from "src/L2/OptimismSuperchainERC20.sol";
+import { SafeCall } from "src/libraries/SafeCall.sol";
+
+contract MockCrossDomainMessenger {
```
Is not the cross domain but the L2 <-> Messenger instead. I suggest also changing the name to the file

```suggestion
contract MockL2ToL2Messenger {
```

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol:
```diff
@@ -0,0 +1,180 @@
+        }
+    }
+
+    function fuzz_MintSupertoken(uint256 index, uint96 amount) external {
```
Missing NATSPEC?

0xDiscotech replied:
I think a good idea could be to split the function to have the properties test functions on one place, and the test-helpers functions in other.
Wdyt?

0xteddybear replied:
good idea!

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol:
```diff
@@ -0,0 +1,180 @@
+    /// @custom:property sum of supertoken total supply across all chains is always equal to convert(legacy, super)-
+    /// convert(super, legacy)
+    /// @dev deliberately not a view method so medusa runs it but not the view methods defined by Test
+    function property_totalSupplyAcrossChainsEqualsMints() external {
```
Maybe this?
```suggestion
    function fuzz_totalSupplyAcrossChainsEqualsMints() external {
```

0xteddybear replied:
this is the only method so far where we manage to _only_ do invariant checking :D

I believe it's worthwhile to give them different names, so later it's more clear if it makes sense to have separate contracts for state transitions or assertions, similar to horsefacts/WETH9

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol:
```diff
@@ -0,0 +1,180 @@
+    /// convert(super, legacy)
+    /// @dev deliberately not a view method so medusa runs it but not the view methods defined by Test
+    function property_totalSupplyAcrossChainsEqualsMints() external {
+        for (uint256 i = 0; i < ghost_totalSupplyAcrossChains.length(); i++) {
+            uint256 totalSupply = 0;
+            (bytes32 currentSalt, uint256 trackedSupply) = ghost_totalSupplyAcrossChains.at(i);
+            for (uint256 j = 0; j < MAX_CHAINS; j++) {
+                address supertoken = MESSENGER.superTokenAddresses(j, currentSalt);
+                if (supertoken != address(0)) {
+                    totalSupply += OptimismSuperchainERC20(supertoken).totalSupply();
```
Could you add some comments or use another var names instead of `_i` of `_j`? bc is kinda difficult to understand hehe

0xDiscotech reviewed (commented):
Astonishing 🫂 🐻

0xteddybear committed:
fix: fixes after lovely feedback by disco

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| id  | description                                                                        | halmos | medusa |
+| --- | ---                                                                                | ---    | ---    |
+| 0   | supertoken token address does not depend on the executing chain’s chainID          | [ ]    | [ ]    |
+| 1   | supertoken token address depends on name, remote token, address and decimals       | [ ]    | [ ]    |
```
you mean symbol instead of address, right?

0xteddybear replied:
:see_no_evil:

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+# supertoken properties
```
you include invariants from liquidity migration. Wdyt about renaming this title?

0xParti replied:
*and factory

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| --- | ---                                                                                | ---    | ---    |
+| 0   | supertoken token address does not depend on the executing chain’s chainID          | [ ]    | [ ]    |
+| 1   | supertoken token address depends on name, remote token, address and decimals       | [ ]    | [ ]    |
+| 2   | convert() should only allow converting legacy tokens to supertoken and viceversa   | [ ]    | [ ]    |
```
allowed legacy and allowed supertokens

0xParti replied:
not every legacy nor supertoken will be allowed

0xteddybear replied:
this was defined in `SUMMARY.md`, however I realize it's not really clear from this document, so I merged them both and made sure the definitions are above

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| 2   | convert() should only allow converting legacy tokens to supertoken and viceversa   | [ ]    | [ ]    |
+| 3   | convert() only allows migrations between tokens representing the same remote asset | [ ]    | [ ]    |
+| 4   | convert() only allows migrations from tokens with the same decimals                | [ ]    | [ ]    |
+| 5   | convert() burns the same amount of one token that it mints of the other            | [ ]    | [ ]    |
```
this sentence is incomplete

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| 12  | supertoken total supply only increases on calls to mint() by the L2toL2StandardBridge             | [ ]    | [ ]    |
+| 13  | supertoken total supply only decreases on calls to burn() by the L2toL2StandardBridge             | [ ]    | [ ]    |
+| 14  | supertoken total supply starts at zero                                                            | [ ]    | [ ]    |
+| 15  | deploying a supertoken registers its remote token in the factory                                  | [ ]    | [ ]    |
```
if you are adding invariants for the factory, you should also add uints for it, right?

0xteddybear replied:
I read through the preliminar implementations of the factories and didn't find anything that struck me as needing an extra layer of testing on top of foundry unit tests, do you have anything in mind?

0xParti replied:
mm good point. Not sure about this, but some ideas:
- Something around metadata initialization? what happens if implementation changes and adds a state variable. How should reinitialization work? We cannot overwrite to the same storage where metadata lives.
- There is only one Superchain token for a fixed set of metadata (this should be obvious from CREATE2 address deduction).
- Every Beacon Proxy deployed must implement the implementation that lives in the address indicated by the Beacon Contract (obvious from the implementation).

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| --- | ---                                                                                                                                                                   | ---    | ---    |
+| 17  | only calls to convert(legacy, super) can increase a supertoken’s  total supply across chains                                                                          | [ ]    | [ ]    |
+| 18  | only calls to convert(super, legacy) can decrease a supertoken’s  total supply across chains                                                                          | [ ]    | [ ]    |
+| 19  | sum of total supply across all chains is always <= to convert(legacy, super)- convert(super, legacy)                                                                  | [ ]    | [ ]    |
```
total supply of what token?

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| 18  | only calls to convert(super, legacy) can decrease a supertoken’s  total supply across chains                                                                          | [ ]    | [ ]    |
+| 19  | sum of total supply across all chains is always <= to convert(legacy, super)- convert(super, legacy)                                                                  | [ ]    | [ ]    |
+| 20  | tokens sendERC20-ed on a source chain to a destination chain can be relayERC20-ed on it as long as the source chain is in the dependency set of the destination chain | [ ]    | [ ]    |
+| 21  | sum of supertoken total supply across all chains is = to convert(legacy, super)- convert(super, legacy) when all cross-chain messages are processed                   | [ ]    | [ ]    |
```
<3

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+| --- | ---                                                                                                                                | ---    | ---     |
+| 22  | sendERC20 decreases sender balance in source chain and increases receiver balance in destination chain exactly by the input amount | [ ]    | [x]     |
+| 23  | sendERC20 decreases total supply in source chain and increases it in destination chain exactly by the input amount                 | [ ]    | [x]     |
+| 24  | sum of supertoken total supply across all chains is always equal to convert(legacy, super)- convert(super, legacy)                 | [ ]    | [~]     |
```
this repeats 19, or am I missing something?

0xteddybear replied:
the property, from the point of view of the contracts under test, is the same one, but it will be instrumented differently (using an atomic bridge), so instead of being equal _when_ all cross chain messages are processed, it must _always_  be equal

If you believe the distinction should be an implementation detail of the testing contracts and readers of this document won't care much, I'm open to removing it

0xParti replied:
gotcha <3

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -0,0 +1,72 @@
+- `[~]`: partially tested/proven property
+- `:(`: property won't be tested due to some limitation
+
+## Unit test
```
you might want to include Event emissions here

0xteddybear replied:
While I recognize the importance of emitting the events for the interop bridge, I don't see how unit tests for that in Foundry would not be sufficient, since the emission or params of it don't depend on state that medusa could push into inconsistency for example

(and I'm not sure the medusa/halmos executing environment can actually assert on event emmissions :sweat_smile:)

0xParti replied:
you are right

0xParti reviewed (commented):
It looks good overall. I would like to see the invariants divided (or at least labeled) by contract/functionality, besides type. There are four things here: the deployments of `OptimismSuperchainERC20` and the`OptimismSuperchainERC20Factory`; and modifications to `L2StandardBridge` (convert) and `OptimismMintableERC20Factory`. It's all kinda mixed

0xteddybear committed:
docs: merge both documents and categorized properties by their milestone

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -1,4 +1,57 @@
+## Definitions
+
+- *legacy token:*  an OptimismMintableERC20 or L2StandardERC20 token on the suprechain that has either been deployed by the factory after the liquidity migration upgrade to the latter, or has been deployed before it **but** added to factory’s `deployments` mapping as part of the upgrade. This testing campaign is not concerned with tokens on L1 or not listed in the factory’s `deployments` mapping.
+- *supertoken:* a SuperchainERC20 contract deployed on the Superchain
```
I would change this: deployed by the `OptimismSuperchainERC20Factory`. Deployed in the superchain is not enough

0xteddybear replied:
updated that + made sure we always refer to said contract by that name

0xParti commented on packages/contracts-bedrock/test/properties/PROPERTIES.md:
```diff
@@ -9,56 +62,56 @@ legend:
-| 5   | convert() burns the same amount of one token that it mints of the other            | [ ]    | [ ]    |
+| id  | milestone           | description                                                                        | halmos | medusa |
+| --- | ---                 | ---                                                                                | ---    | ---    |
+| 0   | Factories           | supertoken token address does not depend on the executing chain’s chainID          | [ ]    | [ ]    |
```
love the new labels :)

0xteddybear committed:
fix: fixes from parti's review

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol:
```diff
@@ -0,0 +1,199 @@
+        remoteTokens.push(address(uint160(1000 + remoteTokens.length)));
+    }
+
+    /// @custom:property-id 14
+    /// @custom:property supertoken total supply starts at zero
```
Ah got it. That's why you didn't use property natspec on `fuzz_DeployNewSupertoken`.
Isn't better to just declare always the property natspec on the test rather than on the function?

0xteddybear replied:
I wanted to define it closest to where it's asserted so there's less of a chance it becomes outdated. Although it's true that we usually want to know which properties are asserted on by a particular _test_

0xDiscotech commented on packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol:
```diff
@@ -0,0 +1,199 @@
+        destinationChainId = bound(destinationChainId, 0, MAX_CHAINS - 1);
+        fromIndex = bound(fromIndex, 0, allSuperTokens.length - 1);
+        OptimismSuperchainERC20 sourceToken = OptimismSuperchainERC20(allSuperTokens[fromIndex]);
+        OptimismSuperchainERC20 destinationToken =
```
Don't get why the destination token address depends also on the `msg.sender`

0xDiscotech replied:
And also, I think in this way you're limiting the test to check balances the same address on origin and destination (`msg.sender`).
IMO it should be more exhaustive if we check on different addresses as well, but it's ok if that adds even more complexity and isn't worth it

0xteddybear replied:
In this PR I tried to get medusa running for the simplest possible (but still somewhat meaningful) test. I have #22 where I add cross-chain transfers between different accounts, which I deemed out of scope for this one since it requires an actor setup and this PR is big enough already :grin:

0xDiscotech reviewed (commented):
Looking good so far ser. The assertions on the tests are correct.
But I find it quite difficult to follow and understand, mostly the setup part. Of course, this is not your fault since the system is complex by itself and we need to a multichain scenario.
But in order to improve it, could we:

* Couple helpers-setup functions and then property functions togheter?
* Divide it into a handler with actors and then another contract file just for test? (maybe is planned but out of scope for this PR)

And also, I have one specific question:
* Where are the supertoken functions being exposed to the fuzzer? I only see that you call a specific function on each test, but how can the fuzzer perform a sequence of calls before it.

0xteddybear committed:
fix: feedback from disco

0xteddybear committed:
fix: feedback from doc

0xteddybear committed:
refactor: separate state transitions from pure properties

0xteddybear committed:
docs: update tested properties

0xteddybear committed:
refactor: move all assertions into properties contract

0xteddybear committed:
fix: move function without assertions back into handler

0xteddybear committed:
test: only use assertion mode

0xteddybear committed:
fix: improve justfile recipie for medusa

simon-something reviewed (approved):
lgtm
